File: android/app/src/main/java/top/itning/yunshu/music/yunshu_music/service/MusicPlayMode.java
package top.itning.yunshu.music.yunshu_music.service;

import android.support.v4.media.session.PlaybackStateCompat;

/**
 * @author itning
 * @since 2021/10/12 15:05
 */
public enum MusicPlayMode {
    SEQUENCE,
    RANDOM,
    LOOP,
    ;

    public static MusicPlayMode fromRepeatMode(@PlaybackStateCompat.RepeatMode int repeatMode) {
        switch (repeatMode) {
            case PlaybackStateCompat.REPEAT_MODE_ONE:
                return MusicPlayMode.LOOP;
            case PlaybackStateCompat.REPEAT_MODE_INVALID:
            case PlaybackStateCompat.REPEAT_MODE_NONE:
            case PlaybackStateCompat.REPEAT_MODE_ALL:
            case PlaybackStateCompat.REPEAT_MODE_GROUP:
            default:
                return MusicPlayMode.SEQUENCE;
        }
    }

    public static MusicPlayMode fromShuffleMode(@PlaybackStateCompat.ShuffleMode int shuffleMode) {
        switch (shuffleMode) {
            case PlaybackStateCompat.SHUFFLE_MODE_ALL:
            case PlaybackStateCompat.SHUFFLE_MODE_GROUP:
                return MusicPlayMode.RANDOM;
            case PlaybackStateCompat.SHUFFLE_MODE_INVALID:
            case PlaybackStateCompat.SHUFFLE_MODE_NONE:
            default:
                return MusicPlayMode.SEQUENCE;
        }
    }

    public static MusicPlayMode getNext(MusicPlayMode nowMode) {
        switch (nowMode) {
            case SEQUENCE:
                return MusicPlayMode.RANDOM;
            case RANDOM:
                return MusicPlayMode.LOOP;
            case LOOP:
            default:
                return MusicPlayMode.SEQUENCE;
        }
    }
}

File: music_platform_interface/lib/music_play_mode.dart
enum MusicPlayMode { SEQUENCE, RANDOMLY, LOOP }

extension MusicPlayModeExtension on MusicPlayMode {
  String name() {
    switch (index) {
      case 0:
        return 'SEQUENCE';
      case 1:
        return 'RANDOMLY';
      case 2:
        return 'LOOP';
      default:
        return 'SEQUENCE';
    }
  }
}

MusicPlayMode valueOf(String name) {
  switch (name) {
    case 'SEQUENCE':
      return MusicPlayMode.SEQUENCE;
    case 'RANDOMLY':
      return MusicPlayMode.RANDOMLY;
    case 'LOOP':
      return MusicPlayMode.LOOP;
    default:
      return MusicPlayMode.SEQUENCE;
  }
}
